### PR TITLE
Stabilize armv8 neon instruction set on aarch64

### PR DESCRIPTION
**Pull request by SparrowLii**
According to the description in [#90972](https://github.com/rust-lang/rust/issues/90972) and the fcp has entered the final comment period stage, I think we can stabilize the armv8 neon instruction set on aarch64.

### CODE REVIEW

File: crates/core_arch/src/aarch64/crc.rs [hunk 1]
```diff
@@ -13,6 +13,7 @@ use stdarch_test::assert_instr;
 #[inline]
 #[target_feature(enable = "crc")]
 #[cfg_attr(test, assert_instr(crc32x))]
+#[stable(feature = "neon_intrinsics", since = "1.59.0")]
 pub unsafe fn __crc32d(crc: u32, data: u64) -> u32 {
```
Comment (Amanieu) [inline crates/core_arch/src/aarch64/crc.rs:17]: The FCP only covers the NEON (SIMD) intrinsics, not crc intrinsics.

File: crates/core_arch/src/aarch64/v8.rs [hunk 2]
```diff
@@ -11,20 +11,23 @@ use stdarch_test::assert_instr;
 /// Reverse the order of the bytes.
 #[inline]
 #[cfg_attr(test, assert_instr(rev))]
+#[stable(feature = "neon_intrinsics", since = "1.59.0")]
 pub unsafe fn _rev_u64(x: u64) -> u64 {
```
Comment (Amanieu) [inline crates/core_arch/src/aarch64/v8.rs:15]: Same here, there intrinsics are not covered by the FCP.

File: crates/core_arch/src/aarch64/tme.rs [hunk 3]
```diff
@@ -71,7 +71,7 @@ pub const _TMFAILURE_TRIVIAL: u64 = 1 << 24;
 /// [ARM TME Intrinsics](https://developer.arm.com/docs/101028/0010/transactional-memory-extension-tme-intrinsics).
 #[inline]
 #[target_feature(enable = "tme")]
-#[cfg_attr(test, assert_instr(tstart))]
+#[cfg_attr(test, assert_instr(nop))]
```
Comment (Amanieu) [inline crates/core_arch/src/aarch64/tme.rs:74]: Why were these changed to nop? This is incorrect.

File: crates/stdarch-gen/neon.spec [hunk 4]
```diff
@@ -7044,7 +7044,7 @@ a = 1.1, 1.9, -1.7, -2.3
 validate 1.0, 2.0, -2.0, -2.0
 target = frintts
 
-aarch64 = frint32x
+aarch64 = nop
```
Comment (Amanieu) [inline crates/stdarch-gen/neon.spec:7047]: Why is this changed to nop? This is incorrect.

File: crates/core_arch/src/arm_shared/neon/mod.rs [hunk 5]
```diff
@@ -4811,7 +6687,7 @@ pub unsafe fn vpadalq_u32(a: uint64x2_t, b: uint32x4_t) -> uint64x2_t {
 #[target_feature(enable = "neon")]
 #[cfg_attr(target_arch = "arm", target_feature(enable = "v8"))]
 #[cfg_attr(all(test, target_arch = "arm"), assert_instr(nop))]
-#[cfg_attr(all(test, target_arch = "aarch64"), assert_instr(smmla))]
+#[cfg_attr(all(test, target_arch = "aarch64"), assert_instr(nop))]
```
Comment (Amanieu) [inline crates/core_arch/src/arm_shared/neon/mod.rs:6690]: Why is this changed to nop? This is incorrect.

Also why is this function not marked as stable?

File: crates/stdarch-gen/src/main.rs [hunk 6]
```diff
@@ -1064,9 +1064,9 @@ fn gen_aarch64(
     };
     let current_target = match target {
         Default => "neon",
-        ArmV7 => "v7",
-        Vfp4 => "vfp4",
-        FPArmV8 => "fp-armv8,v8",
+        ArmV7 => "neon",
+        Vfp4 => "neon",
+        FPArmV8 => "neon",
```
Comment (Amanieu) [inline crates/stdarch-gen/src/main.rs:1069]: Why did you change this? The previous target features were correct.

**Comment by Amanieu** [comment]
We still need to wait ~1 week for the FCP to complete. Once it is complete, this PR can be merged.

File: crates/core_arch/src/aarch64/crc.rs (hunk 1 above)
Comment (SparrowLii) [reply crates/core_arch/src/aarch64/crc.rs]: OK. Unmarked it.

File: crates/core_arch/src/aarch64/v8.rs (hunk 2 above)
Comment (SparrowLii) [reply crates/core_arch/src/aarch64/v8.rs]: OK. Unmarked it.

File: crates/core_arch/src/aarch64/tme.rs (hunk 3 above)
Comment (SparrowLii) [reply crates/core_arch/src/aarch64/tme.rs]: The `assert_instr` of these instructions failed on my local aarch64. But since these instructions are not included in this stabilization, I think we can keep them unchanged.

File: crates/stdarch-gen/neon.spec (hunk 4 above)
Comment (SparrowLii) [reply crates/stdarch-gen/neon.spec]: Same as above. Changes have been cancelled.

File: crates/core_arch/src/arm_shared/neon/mod.rs (hunk 5 above)
Comment (SparrowLii) [reply crates/core_arch/src/arm_shared/neon/mod.rs]: Same as above. And vmmla instructions require `v8.6a` architecture, so they are not included in this stabilization.

File: crates/stdarch-gen/src/main.rs (hunk 6 above)
Comment (SparrowLii) [reply crates/stdarch-gen/src/main.rs:1069]: There are two functions used to generate instructions: `gen_arm` and `gen_aarch64`. The part I changed is in `gen_aarch64`. The values of these targets do not affect the `target_feature` on aarch64, so unified to `neon` I think it is more appropriate

File: crates/core_arch/src/arm_shared/neon/mod.rs [hunk 7]
```diff
@@ -4179,6 +5835,10 @@ pub unsafe fn vext_s64<const N: i32>(a: int64x1_t, _b: int64x1_t) -> int64x1_t {
 #[cfg_attr(all(test, target_arch = "arm"), assert_instr("nop", N = 0))]
 #[cfg_attr(all(test, target_arch = "aarch64"), assert_instr("nop", N = 0))]
 #[rustc_legacy_const_generics(2)]
+#[cfg_attr(
+    target_arch = "aarch64",
+    stable(feature = "neon_intrinsics", since = "1.59.0")
+)]
 pub unsafe fn vext_u64<const N: i32>(a: uint64x1_t, _b: uint64x1_t) -> uint64x1_t {
     if N != 0 {
         unreachable_unchecked()
```
Comment (Amanieu) [inline crates/core_arch/src/arm_shared/neon/mod.rs:5844]: I just noticed this: we should be using a static assert here, not `unreachable_unchecked`. This applies to all `vext_{s,u,p,f}64`.

File: crates/stdarch-gen/src/main.rs (hunk 6 above)
Comment (Amanieu) [reply crates/stdarch-gen/src/main.rs:1069]: That's fine if it's only on aarch64.

File: crates/core_arch/src/aarch64/tme.rs (hunk 3 above)
Comment (Amanieu) [reply crates/core_arch/src/aarch64/tme.rs]: It might have failed because the binutils version on your system is too old and objdump is not able to disassemble the instruction. We use a recent version of ubuntu in the docker image used in CI for this reason.

**Comment by glandium** [comment]
Is it an oversight that many of these now stable functions take arguments with types that still aren't stable? (I say many, but I haven't checked, it could just as well be all of them)

**Comment by thomcc** [comment]
It’s not an oversight, it’s a rustdoc bug. Those are stable on aarch64, but not arm32, so it gets mislabeled (theres an open issue for it, I can’t provide it at the moment though)
